src/crypto/CHIPCryptoPAL.cpp

```diff
@@ -648,5 +649,185 @@ CHIP_ERROR GenerateCompressedFabricId(const Crypto::P256PublicKey & root_public_
+
+    ReturnErrorOnFailure(hashStream.Begin());
+    ReturnErrorOnFailure(hashStream.AddData(attestationElements));
+    ReturnErrorOnFailure(hashStream.AddData(attestationChallenge));
+    ReturnErrorOnFailure(hashStream.Finish(messageDigestSpan));
+
+    ReturnErrorOnFailure(ECDSA_sign_hash(messageDigestSpan.data(), messageDigestSpan.size(), out_signature));
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR P256PublicKey::ECDSA_validate_attestation_data(const ByteSpan & attestationElements,
+                                                          const ByteSpan & attestationChallenge,
+                                                          const P256ECDSASignature & out_signature) const
+{
+    Hash_SHA256_stream hashStream;
+    uint8_t md[kSHA256_Hash_Length];
+    MutableByteSpan messageDigestSpan(md);
+
+    ReturnErrorOnFailure(hashStream.Begin());
+    ReturnErrorOnFailure(hashStream.AddData(attestationElements));
+    ReturnErrorOnFailure(hashStream.AddData(attestationChallenge));
+    ReturnErrorOnFailure(hashStream.Finish(messageDigestSpan));
+
+    ReturnErrorOnFailure(ECDSA_validate_hash_signature(messageDigestSpan.data(), messageDigestSpan.size(), out_signature));
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ConstructAttestationElements(const ByteSpan & certificationDeclaration, const ByteSpan & attestationNonce,
+                                        uint32_t timestamp, const ByteSpan & firmwareInfo, const ByteSpan & vendorReserved5,
+                                        const ByteSpan & vendorReserved6, const ByteSpan & vendorReserved7,
+                                        const ByteSpan & vendorReserved8, MutableByteSpan & attestationElements)
+{
+    TLV::TLVWriter tlvWriter;
+    TLV::TLVType outerContainerType = TLV::kTLVType_NotSpecified;
+
+    tlvWriter.Init(attestationElements.data(), static_cast<uint32_t>(attestationElements.size()));
+    outerContainerType = TLV::kTLVType_NotSpecified;
+    ReturnErrorOnFailure(tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(1), certificationDeclaration));
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(2), attestationNonce));
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(3), timestamp));
+    if (!firmwareInfo.empty())
+    {
+        ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(4), firmwareInfo));
+    }
+    if (!vendorReserved5.empty())
+    {
+        ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(5), vendorReserved5));
+    }
+    if (!vendorReserved6.empty())
+    {
+        ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(6), vendorReserved6));
+    }
+    if (!vendorReserved7.empty())
+    {
+        ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(7), vendorReserved7));
+    }
+    if (!vendorReserved8.empty())
+    {
+        ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(8), vendorReserved8));
+    }
```

Comment:
The scheme for vendor-reserved used here is not spec compliant. Recent changes were made to normalize it, see https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/3598/files